### ISSUE TITLE
API: incorrect declarations in types.ts

### ISSUE THREAD

**Issue by ajafff** [Microsoft/TypeScript#24706]
The `--strictNullChecks` PR changed some declarations in `types.ts` to avoid frequent assertions in the compiler's code. Unfortunately these changes make it really unsafe for API users because they no longer contain the actually nullable types: https://github.com/Microsoft/TypeScript/pull/22088#discussion_r184149465 https://github.com/Microsoft/TypeScript/pull/22088#discussion_r184155027

* `ts.Node#parent` is no longer optional (which is definitely wrong for `SourceFile`)
* `ts.Symbol#declarations` and `ts.Symbol.valueDeclaration` are no longer optional, but can still be undefined
* `ts.Type#symbol` is no longer optional, but I guess this comment is still up to date: `// Symbol associated with type (if any)`

These changes should either be reverted or replaced in the published declaration files.
/cc @andy-ms @weswigham

**Comment by weswigham** [comment]
I agree. I mentioned it on the PR. Now that we've done the bulk of the changes, we should fix one at a time and introduce helpers to help the pain.

**Comment by ajafff** [comment]
@RyanCavanaugh Do you plan to fix this in the near future? It's a real PITA for me as I use TypeScript's API a lot.
Is there anything I can do to help? Would you accept PRs to fix the issues listed above (probably one at a time to keep the diff small)?

**Comment by bradzacher** [comment]
We just ran into this bug in `@typescript-eslint`. One of our rules assumed the typing was correct, causing a bug for an end user.
As an API consumer, we rely on the types to help us ensure we catch bugs at compile time, and it's counter-intuitive to handle cases that are contrary to the types.

**Comment by ayazhafiz** [comment]
@andy-ms on a high level, what would be the changes required to type `Symbol.valueDeclaration` as `ts.Declaration | undefined` in a [`.d.ts` postproccessing step](https://github.com/microsoft/TypeScript/pull/22088#discussion_r184162921)?

**Comment by ayazhafiz** [comment]
@weswigham @RyanCavanaugh will you accept changes fixing this?
